sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -237,21 +237,27 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
           new Path(metastoreRelation.catalogTable.storage.locationUri.get),
           partitionSpec)
 
-        val inferredSchema = if (fileType.equals("parquet")) {
-          val inferredSchema =
-            defaultSource.inferSchema(sparkSession, options, fileCatalog.allFiles())
-          inferredSchema.map { inferred =>
-            ParquetFileFormat.mergeMetastoreParquetSchema(metastoreSchema, inferred)
-          }.getOrElse(metastoreSchema)
-        } else {
-          defaultSource.inferSchema(sparkSession, options, fileCatalog.allFiles()).get
+        val schema = fileType match {
+          case "parquet" =>
+            val inferredSchema =
+              defaultSource.inferSchema(sparkSession, options, fileCatalog.allFiles())
+
+            // For Parquet, get correct schema by merging Metastore schema data types
+            // and Parquet schema field names.
+            inferredSchema.map { schema =>
+              ParquetFileFormat.mergeMetastoreParquetSchema(metastoreSchema, schema)
+            }.getOrElse(metastoreSchema)
+          case "orc" =>
+            metastoreSchema
+          case _ =>
+            throw new RuntimeException(s"Cannot convert a $fileType to a HadoopFsRelation")
```

Comment:
Spark SQL users do not know what is a `HadoopFsRelation`. 

`to a data source table`?